substrate/client/consensus/beefy/src/communication/fisherman.rs

```diff
@@ -0,0 +1,363 @@
+				{
+					Ok(Ok(ancestry_proof)) => Some(ancestry_proof),
+					Ok(Err(e)) => {
+						debug!(target: LOG_TARGET, "🥩 Failed to generate ancestry proof: {:?}", e);
+						None
+					},
+					Err(e) => {
+						debug!(target: LOG_TARGET, "🥩 Failed to generate ancestry proof: {:?}", e);
+						None
+					},
+				};
+				let proof = ForkEquivocationProof {
+					commitment: vote.commitment,
+					signatories: vec![(vote.id, vote.signature)],
+					canonical_header: Some(canonical_header),
+					ancestry_proof,
+				};
+				self.report_fork_equivocation(proof)?;
+			}
+		}
+		Ok(())
+	}
+
+	/// Check `signed_commitment` for contained block against canonical payload. If an equivocation
+	/// is detected, this also reports it.
+	fn check_signed_commitment(
+		&self,
+		signed_commitment: SignedCommitment<NumberFor<B>, Signature>,
+	) -> Result<(), Error> {
+		let SignedCommitment { commitment, signatures } = signed_commitment;
+		let number = commitment.block_number;
+		// if the vote is for a block number exceeding our best block number, there shouldn't even
+		// be a payload to sign yet, hence we assume it is an equivocation and report it
+		if number > self.backend.blockchain().info().best_number {
+			// if block number is in the future, we use the latest validator set
+			// as the assumed signatories (note: this assumption is fragile and can possibly be
+			// improved upon)
+			let best_hash = self.backend.blockchain().info().best_hash;
+			let validator_set = self.active_validator_set_at(best_hash)?;
+			let signatories: Vec<_> = validator_set
+				.validators()
+				.iter()
+				.cloned()
+				.zip(signatures.into_iter())
+				.filter_map(|(id, signature)| match signature {
+					Some(sig) =>
+						if sp_consensus_beefy::check_commitment_signature::<
+							_,
+							_,
+							BeefySignatureHasher,
+						>(&commitment, &id, &sig)
+						{
+							Some((id, sig))
+						} else {
+							None
+						},
+					None => None,
+				})
+				.collect();
+			if signatories.len() > 0 {
+				let proof = ForkEquivocationProof {
+					commitment,
+					signatories,
+					canonical_header: None,
+					ancestry_proof: None,
+				};
+				self.report_fork_equivocation(proof)?;
+			}
+		} else {
+			let (correct_hash, canonical_header, canonical_payload) =
+				self.canonical_hash_header_payload_tuple(number)?;
+			if commitment.payload != canonical_payload {
+				let ancestry_proof = match self.runtime.runtime_api().generate_ancestry_proof(
+					correct_hash,
+					number,
+					None,
+				) {
+					Ok(Ok(ancestry_proof)) => Some(ancestry_proof),
+					Ok(Err(e)) => {
+						debug!(target: LOG_TARGET, "🥩 Failed to generate ancestry proof: {:?}", e);
+						None
+					},
+					Err(e) => {
+						debug!(target: LOG_TARGET, "🥩 Failed to generate ancestry proof: {:?}", e);
+						None
+					},
+				};
+				let validator_set = self.active_validator_set_at(correct_hash)?;
+				if signatures.len() != validator_set.validators().len() {
```

Comment:
It's correct that a valid `SignedCommitment` will have between `[2/3+1, 3/3]` of `validator_set.validators().len()` entries with `Some(sig)`.
But `signatures.len()` is nonetheless strictly equal to `validator_set.validators().len()`, so the guardian clause here is valid (else, checks like [this](https://github.com/lederstrumpf/polkadot-sdk/blob/d27fc06a279f40f18c5255f48559c199c9b0a3a6/substrate/client/consensus/beefy/src/communication/fisherman.rs#L358-L378) wouldn't work).

We'd else have to turn `SignedCommitment` into a map of validator IDs to sigs, and for now that seems like a worse tradeoff (since these need to be parsed on the BEEFY light client too) than up to a third of `None` entries.

> You should use helper [verify_with_validator_set(number, validator_set, proof)](https://github.com/paritytech/polkadot-sdk/blob/f2f4b154d7641ef3c237a0d7df4b77280b5a81d0/substrate/client/consensus/beefy/src/justification.rs#L52) to avoid duplicating these checks.

Good point - done in https://github.com/paritytech/polkadot-sdk/pull/1903/commits/d27fc06a279f40f18c5255f48559c199c9b0a3a6